### ISSUE TITLE
BaseError hack + closure-compiler hack = constructor error

### ISSUE THREAD

**Issue by brad4d** [angular/angular#13864]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

An exception something like this one occurs when constructing a class that inherits from `BaseError` in code compiled with closure-compiler.

```
Uncaught TypeError: Cannot set property 'message' of undefined
    at module$contents$$40angular$core$src$di$reflective_errors_NoAnnotationError.set [as message] (.../app_bundle.js:4637:19)
    at module$contents$$40angular$core$src$di$reflective_errors_NoAnnotationError.module$contents$$40angular$core$src$facade$errors_BaseError [as constructor]
    at new module$contents$$40angular$core$src$di$reflective_errors_NoAnnotationError
    at 
```

This occurs because closure-compiler tries to make inheritance from `Error` work by transpiling the
`BaseError` constructor from something like this

```javascript
const BaseError = class extends Error {
  constructor() {
    super(message);
    // Errors don't use current this, instead they create a new instance.
    // We have to do forward all of our api to the nativeInstance.
    // TODO(bradfordcsmith): Remove this hack when
    //     google/closure-compiler/issues/2102 is fixed.
    const nativeError = new Error(message) as any as Error;
    this._nativeError = nativeError;  }
...
}
```
to this
```javascript
var BaseError = function(message) {
  var $jscomp$tmp$error;
  $jscomp$tmp$error = Error.call(this, message);
  this.message = $jscomp$tmp$error.message;
  "stack" in $jscomp$tmp$error && (this.stack = $jscomp$tmp$error.stack);
  this.Mb = Error(message);
};
```
The intention is to deal with the fact that the `Error`  constructor *always* returns a new object and ignores the `this` you pass to it by just copying over the two most important fields.
(Copying of `stack` is further complicated because old versions of IE actually don't fill it in until the
`Error` object is thrown and won't do it then if the object has a `stack` field already, so we must actually *avoid* copying it if it is undefined.)

Trouble is the `BaseError` code also has its own code that tries to work around this inheritance problem
by delegating all references to known `Error` fields to a private `Error` object.
```javascript
// within the BaseError class definition
  get message() { return this._nativeError.message; }
  set message(message) { this._nativeError.message = message; }
  get name() { return this._nativeError.name; }
  get stack() { return (this._nativeError as any).stack; }
  set stack(value) { (this._nativeError as any).stack = value; }
  toString() { return this._nativeError.toString(); }
```

This causes the statement that assigns to `message` in the constructor to fail because the constructor hasn't assigned the private `_nativeError` field (renamed to `Mb` in the example code) yet.

**Expected behavior**
Since angular code isn't always compiled with closure-compiler, it cannot rely on
the closure-compiler hack to make inheritance from Error to work correctly.
Unfortunately, its current solution conflicts with closure-compiler's solution.

Is it really necessary for angular to use getters and setters, though?
I think that's more cleverness than is required.

1. message just needs to be copied during construction.
2. stack is a bit trickier as described above, but there's no need for delegation.
3. name is on the prototype, there's really no reason to copy it.
4. toString() will already be inherited from the Error prototype, so there's no need to delegate it either.

I'd like to suggest that the BaseError class be changed to this.

```javascript
export class BaseError extends Error {
  constructor(message: string) {
    const superThis = super(message) as any as Error;
    // When this code is transpiled to ECMAScript 5 - compliant code the super() call may not work
    // properly when extending Error classes. It depends on what program is doing the transpilation.
    // Typically super(message) becomes Error.call(this, message). Unfortunately, Error.call() ignores the
    // `this` value passed to it and always returns a new Error object. If this code is not transpiled
    // or is transpiled correctly, then super() will return `this`.
    // If not, then super() will return a new Error object from which we need to copy fields.
    if (superThis != this) {
      this.message = superThis.message;
      if (superThis.stack) {
        // Workaround for IE11 and earlier which don't set stack until the object is thrown and won't
        // set it at all if the field already exists. Later browsers set stack during object construction.
        this.stack = superThis.stack;
      }
    }
  }
}

```
**Minimal reproduction of the problem with instructions**
Compile and a minimal code sample that creates a BaseError.

**What is the motivation / use case for changing the behavior?**
Make BaseError code work correctly
- when transpiled by closure-compiler
- when untranspiled
- when transpiled by programs that don't have special handling for inheritance from Error.

**Please tell us about your environment:**
All

* **Angular version:** 2.0.X
Any with the version of BaseError described above.

* **Browser:** [all]
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript or ES6/7 transpiled to ES5 by closure-compiler

* **Node (for AoT issues):** `node --version` =

**Comment by brad4d** [comment]
FYI, I've just confirmed that `name` is on `Error.prototype` even in IE8.
So, there's really no need to have a getter for it.

**Comment by brad4d** [comment]
This is closely related to #12575 
Which includes a reference to a TypeScript update regarding how `super()` is handled.
https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work

It seems that in the latest versions of TypeScript the decision was made to just go ahead and let `super()` always replace the value of `this` with the result of `SuperClass.call(this, ....)`,
which means the prototype chain is broken when inheriting from `Error`.
The recommended fix doesn't work for IE <= 10.

So we have, I think, 3 cases that must work.

1.  TypeScript transpiles ES6 classes -> ES5 classes.
    In this case the prototype chain will be broken. The current code works for this case by delegating
    fields to a private Error object, but `instanceof` is broken.
    My suggested fix in the description of this issue won't work correctly, because I was assuming the prototype chain was correct.
2.  TypeScript transpiles to ES6-compatible code (classes left unchanged), which is then run through closure-compiler.
    In this case we have the problem described in this bug.
    The prototype chain is correct, but the delegation will be broken.
3.  TypeScript transpiles to ES6-compatible code, which is run in an ES6-capable environment.
    We have to avoid breaking this case with the workarounds for cases 1 and 2.

**Comment by brad4d** [comment]
Upon further reflection, actually my original suggestion will be fine.

In all cases where the value returned by `super()` === `this` there's no real need for a workaround.
1. TypeScript -> native ES6 running in ES6 environment (no closure-compiler)
    Inheritance from Error just works.
2. TypeScript -> ES5 (no closure-compiler)
    The constructor returns an object that uses `Error` rather than `BaseError` as its prototype.
    The object will be initialized correctly (message, stack, etc.) by the `Error` constructor.
    Any fields on BaseError.prototype won't be available, but that's a separate issue from what I'm considering here.
3. TypeScript -> ES6 -> closure-compiler -> ES5
    closure-compiler ensures that the object will be initialized correctly as if the `Error` constructor
    had executed on it.

There's only one case where `super()` will return a value that is not equal to `this`.
That's when using an older version of the TypeScript compiler to transpile to ES5-compatible code.
This is the case the current code was designed to handle.
My suggested fix will preserve this workaround.

**Comment by brad4d** [comment]
@mprobst 
Since the current workaround code is only relevant when doing TS -> ES5 with an outdated version of TypeScript, would it be reasonable to just remove it all?

**Comment by mprobst** [comment]
@brad4d sorry, I don't follow the whole issue much. @mhevery seems to know better what's going on.

What's the older TypeScript compiler you're referring to? Angular needs to support 1.8 for the time being, but will move to 2.0 soon (or has already moved?).

**Comment by hlhicks** [comment]
@brad4d, just wanted to add my up-vote for your initial solution: 1) remove property delegates, 2) detect mismatch in `super` return value and copy properties.

I've got two additional thoughts...
1) Typescript specification for `super` [does not mention](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#832-super-calls) a return value, so it would be good to make allowances for an undefined value of superThis. To do so, we could change the detection code: from 
`if (superThis != this) {...}` to `if (superThis && superThis != this) {...}`

2) In practice, the `Error` object is actually a `ZoneAwareError` [(link)](https://github.com/angular/zone.js/blob/c00abdbfceec5ce87fe901fbd207cb4591b48f1a/dist/zone.js#L672).   Doesn't this insulate us from the problems with IE11, et al.   I think removing all the work-around code completely--as you suggested--would be a better solution at this point.

**Comment by brad4d** [comment]
@mprobst, the TypeScript release notes that describe the change to the behavior of `super()` I described above are for TypeScript 2.1.
https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#extending-built-ins-like-error-array-and-map-may-no-longer-work

So, that sounds pretty-darned-new and I don't suppose I should count on it being available.

**Comment by brad4d** [comment]
@hlhicks, thanks for your endorsement.
I'm unsure whether your endorsement is sufficient for me to spend the time creating a pull request.
@mhevery, what do you think?

**Comment by mhevery** [comment]
/cc @JiaLiPassion  

This should now be fixed in Zone v0.7.6 Closing the issue. Please re-open if that is not the case.
